File: Week_03/subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # 迭代解法，每次往已有结果中加入新元素，构成新的子集，然后跟原有子集合并
        # res = [[]]
        # for num in nums:
        #     res += [[num] + item for item in res]
        # return res

        # 回溯解法
        def backtrack(first = 0, curr = []):
            # terminator
            if len(curr) == k:
                # use curr[:] instead of curr
                output.append(curr[:])
            for i in range(first, n):
                # process current level logic
                curr.append(nums[i])
                # drill down
                backtrack(i + 1, curr)
                # reverse state
                curr.pop()

        output = []
        n = len(nums)
        for k in range(n + 1):
            backtrack()
        return output

File: Week_03/subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        // 需要显示初始化为空
        vector<vector<int>> res = {{}};
        if (nums.size() == 0)   return res;
        for (int num : nums) {
            int size = res.size();
            for (int i = 0; i < size; i++) {
                // 先把当前遍历的元素加入 vector 尾部，然后对齐进行拓展
                res.push_back(res[i]);
                res.back().push_back(num);
            }
        }
        return res;
    }
};
